File: leetcode/p00105/p00105.py
import enum
from typing import Optional, List

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(n)
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        self.rootIdx, memo = 0, {val: idx for idx, val in enumerate(inorder)}

        def helper(startIdx: int, endIdx: int) -> TreeNode:
            if startIdx > endIdx:
                 return None
            
            root = TreeNode(preorder[self.rootIdx])
            self.rootIdx += 1
            root.left = helper(startIdx, memo[root.val]-1)
            root.right = helper(memo[root.val]+1, endIdx)
            
            return root
        
        return helper(0, len(inorder)-1)

File: leetcode/p00105/p00105.go
package p00105

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Time Complexity: O(n)
// Space Complexity: O(n)
func buildTree(preorder []int, inorder []int) *TreeNode {
	memo := map[int]int{}
	for idx, val := range inorder {
		memo[val] = idx
	}

	root, _ := helper(preorder, inorder, memo, 0, 0, len(preorder)-1)
	return root
}

func helper(preorder, inorder []int, memo map[int]int, rootIdx, startIdx, endIdx int) (*TreeNode, int) {
	if startIdx > endIdx {
		return nil, rootIdx
	}

	root := &TreeNode{Val: preorder[rootIdx]}
	rootIdx++
	root.Left, rootIdx = helper(preorder, inorder, memo, rootIdx, startIdx, memo[root.Val]-1)
	root.Right, rootIdx = helper(preorder, inorder, memo, rootIdx, memo[root.Val]+1, endIdx)

	return root, rootIdx
}
